site/content/blog/2021-01-21-streamlining-developer-access-to-prometheus-and-grafana.md

```diff
@@ -0,0 +1,57 @@
+    - `inbucket` to simplify email testing without routing outside your machine
+    - `minio` to expose an S3-compatible interface to your local disk for for hosting uploaded files
+    - `openldap` to simplify testing alternate modes of authentication (requires access to the enterprise source)
+    - `elasticsearch` to expose an improved search backend for posts and channels (requires access to the enterprise source)
+* linking your server to an automatically detected `mattermost-webapp` directory
+* optionally starting the server in the foreground if `RUN_SERVER_IN_BACKGROUND` is `false`
+* piping the structured server logs through a `logrus` decorator for easier reading
+* invoking the `go` compiler to run the server code in your local repository
+
+### Adding Prometheus & Grafana
+
+As of https://github.com/mattermost/mattermost-server/pull/16649, this build tooling now supports automating the setup of Prometheus and Grafana Docker containers. Together, these tools form the backbone of our [Performance Monitoring (E20)](https://docs.mattermost.com/deployment/metrics.html) functionality, scraping metrics from Mattermost to help give enterprise customers insight into the performance of their clusters. Automating this setup makes it easier for developers to test in a production-like environment, and to help extend the monitoring available to customers.
+
+Since the corresponding Mattermost functionality is only available with access to the enterprise source code, these containers aren't enabled by default. To turn them on, export the following environment variable into your shell:
+```
+ENABLED_DOCKER_SERVICES = mysql postgres inbucket prometheus grafana
+```
+
+Alternatively, you can follow the instructions in [config.mk](https://github.com/mattermost/mattermost-server/blob/master/config.mk). (Note that it's not necessary to specify `minio`, `openldap` or `elasticsearch` -- these are added automatically as needed.)
+
+The build tooling does more than just spin up two new containers, however. [Various configuration files](https://github.com/mattermost/mattermost-server/tree/master/build/docker) also automate the following:
+* configuring Prometheus to scrape the Mattermost `:8067/metrics` endpoint (requires access to the enterprise source)
+* enabling anonymous access to Grafana (you can still login with the default `admin`/`admin` credentials if needed)
+* configuring Grafana with the containerized Prometheus as the default data source
+* installing our canonical Grafana dashboards ([Mattermost Performance KPI Metrics](https://grafana.com/grafana/dashboards/2539), [Mattermost Performance Monitoring](https://grafana.com/grafana/dashboards/2542), [Mattermost Performance Monitoring (Bonus Metrics)](https://grafana.com/grafana/dashboards/2545))
+
+This means that simply running `make run-server` gets you immediate access to the Mattermost Performance dashboards at `http://localhost:3000`:
```

Comment:
One question, do you have to enable "Metrics" in your mattermost to make it work, right? Probably adding that info here would be great, something like "Set your `MetricsSettings.Enable` to `true` in your configuration file (or through the admin console)".

Reply:
Excellent point! I'll also add a note here about having to restart the server when installing a license -- for some reason, metrics (even if enabled) don't start until the next server start.

Reply:
From my experience it's sufficient to trigger a config save to make the endpoint work. This is assuming the server build is enterprise ready. I think we are missing some logic to start the server when a valid license is added. It might be worth consider fixing this edge case.

Reply:
Indeed! I'll look at that edge cases vs. permanently documenting the pain point here. Otherwise, I've pushed changes to document enabling the metrics. Thanks!